Support IPA IPA Trust with additional IPA server

justin-stephenson opened the pull request:
Add new server `master2.ipa2.test` which deploys an IPA domain `ipa2.test` to be used in IPA IPA trust.

with this PR checked out
`sudo make down`
`sudo make build`
`sudo REGISTRY="localhost/sssd" make up

Linked PRs:
https://github.com/SSSD/sssd-test-framework/pull/119
https://github.com/SSSD/sssd/pull/7517

jakub-vavra-cz commented on src/ansible/roles/dns/templates/etc.dnsmasq.conf.j2:
```diff
@@ -16,6 +16,9 @@ cache-size=0
 {% if 'master.ipa.test' in hostvars %}
 server=/ipa.test/{{ hostvars['master.ipa.test']['ansible_facts']['default_ipv4']['address'] }}
 {% endif %}
+{% if 'master2.ipa2.test' in hostvars %}
```
It would be better to do a loop through all ipa servers.

justin-stephenson replied:
Fixed.

jakub-vavra-cz commented on src/ansible/roles/cleanup/tasks/main.yml:
```diff
@@ -29,4 +29,4 @@
   - name: Remove SSSD's database and logs
     shell: rm -f /var/lib/sss/db/* /var/lib/sss/mc/* /var/log/sssd/*
-  when: inventory_hostname == 'client.test' or inventory_hostname == 'master.ipa.test'
+  when: inventory_hostname == 'client.test' or inventory_hostname == 'master.ipa.test' or inventory_hostname == 'master2.ipa2.test' or inventory_hostname == 'master2.ipa2.test'
```
This should run on group ipa and client.

justin-stephenson replied:
Fixed.

jakub-vavra-cz commented on src/ansible/roles/cleanup/tasks/main.yml:
```diff
@@ -7,7 +7,7 @@
   - name: Remove 389ds database to make image smaller
     shell: rm -f /var/lib/dirsrv/slapd-IPA-TEST/db/__db.*
-  when: inventory_hostname == 'master.ipa.test' or inventory_hostname == 'ipa-devel'
+  when: inventory_hostname == 'master.ipa.test' or inventory_hostname == 'master2.ipa2.test' or inventory_hostname == 'ipa-devel'
```
this should run on group ipa

justin-stephenson replied:
Updated.

jakub-vavra-cz commented on src/ansible/roles/packages/tasks/Fedora.yml:
```diff
@@ -221,7 +221,7 @@
       state: present
       name: sssd-kcm
-  when: "'base_ipa' in group_names or 'ipa' in group_names"
+  when: "'base_ipa' in group_names or 'base_ipa2' in group_names or 'ipa' in group_names"
```
Is there a reason to use base_ipa2? Wouldn't it be tha same as base_ipa?

justin-stephenson replied:
Fixed to use `base_ipa` only, this was an oversight on my part.

jakub-vavra-cz replied:
Does not look fixed yet.

jakub-vavra-cz commented on src/ansible/inventory.yml:
```diff
@@ -14,6 +14,9 @@ all:
         base_ipa:
           hosts:
             base-ipa
+        base_ipa2:
```
why create base_ipa2?

justin-stephenson replied:
Fixed to use base_ipa only, this was an oversight on my part.

justin-stephenson commented:
Hi @pbrezina Can you help me understand why `docker.io/ubuntu:rolling` build fails (https://github.com/SSSD/sssd-ci-containers/actions/runs/10370744252/job/28711693110?pr=106)  when sshd does not restart properly on `master2.ipa2.test`. I see that the host keys are not found.

```
Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.ecdsa_key                                                    
Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.ed25519_key                                          
Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.rsa_key  

```

 I ran `src/tools/gen-ssh-keys.sh` and added the ssh keys for `master2.ipa2.test` into this PR. I also added `master2.ipa2.test` into the for loop in `src/tools/gen-ssh-keys.sh`

However, these are not being copied into the `master2.ipa2.test` system.
~~~
[root@master2 /]# ll data/ssh-keys/hosts/
total 192       
-rw-------. 1 root root  525 Aug 13 02:05 client.test.ecdsa_key                                                        
-rw-------. 1 root root  189 Aug 13 02:05 client.test.ecdsa_key.pub                                                    
-rw-------. 1 root root  419 Aug 13 02:05 client.test.ed25519_key                                                      
-rw-------. 1 root root  109 Aug 13 02:05 client.test.ed25519_key.pub                                                  
-rw-------. 1 root root 2610 Aug 13 02:05 client.test.rsa_key                                                          
-rw-------. 1 root root  581 Aug 13 02:05 client.test.rsa_key.pub                                                      
-rw-------. 1 root root  525 Aug 13 02:05 dc.samba.test.ecdsa_key                                                      
-rw-------. 1 root root  189 Aug 13 02:05 dc.samba.test.ecdsa_key.pub                                                  
-rw-------. 1 root root  419 Aug 13 02:05 dc.samba.test.ed25519_key                                                    
-rw-------. 1 root root  109 Aug 13 02:05 dc.samba.test.ed25519_key.pub                                                
-rw-------. 1 root root 2622 Aug 13 02:05 dc.samba.test.rsa_key                                                        
-rw-------. 1 root root  581 Aug 13 02:05 dc.samba.test.rsa_key.pub                                                    
-rw-------. 1 root root  525 Aug 13 02:05 dns.test.ecdsa_key                                                           
-rw-------. 1 root root  189 Aug 13 02:05 dns.test.ecdsa_key.pub                                                       
-rw-------. 1 root root  419 Aug 13 02:05 dns.test.ed25519_key                                                         
-rw-------. 1 root root  109 Aug 13 02:05 dns.test.ed25519_key.pub                                                     
-rw-------. 1 root root 2610 Aug 13 02:05 dns.test.rsa_key
-rw-------. 1 root root  581 Aug 13 02:05 dns.test.rsa_key.pub                                                         
-rw-------. 1 root root  525 Aug 13 02:05 kdc.test.ecdsa_key                                                           
-rw-------. 1 root root  189 Aug 13 02:05 kdc.test.ecdsa_key.pub                                                       
-rw-------. 1 root root  419 Aug 13 02:05 kdc.test.ed25519_key                                                         
-rw-------. 1 root root  109 Aug 13 02:05 kdc.test.ed25519_key.pub                                                     
-rw-------. 1 root root 2622 Aug 13 02:05 kdc.test.rsa_key
-rw-------. 1 root root  581 Aug 13 02:05 kdc.test.rsa_key.pub                                                         
-rw-------. 1 root root  525 Aug 13 02:05 master.ipa.test.ecdsa_key                                                    
-rw-------. 1 root root  189 Aug 13 02:05 master.ipa.test.ecdsa_key.pub                                                
-rw-------. 1 root root  419 Aug 13 02:05 master.ipa.test.ed25519_key                                                  
-rw-------. 1 root root  109 Aug 13 02:05 master.ipa.test.ed25519_key.pub                                              
-rw-------. 1 root root 2622 Aug 13 02:05 master.ipa.test.rsa_key                                                      
-rw-------. 1 root root  581 Aug 13 02:05 master.ipa.test.rsa_key.pub                                                  
-rw-------. 1 root root  525 Aug 13 02:05 master.keycloak.test.ecdsa_key                                               
-rw-------. 1 root root  189 Aug 13 02:05 master.keycloak.test.ecdsa_key.pub                                           
-rw-------. 1 root root  419 Aug 13 02:05 master.keycloak.test.ed25519_key                                             
-rw-------. 1 root root  109 Aug 13 02:05 master.keycloak.test.ed25519_key.pub                                         
-rw-------. 1 root root 2622 Aug 13 02:05 master.keycloak.test.rsa_key                                                 
-rw-------. 1 root root  581 Aug 13 02:05 master.keycloak.test.rsa_key.pub                                             
-rw-------. 1 root root  525 Aug 13 02:05 master.ldap.test.ecdsa_key                                                   
-rw-------. 1 root root  189 Aug 13 02:05 master.ldap.test.ecdsa_key.pub                                               
-rw-------. 1 root root  419 Aug 13 02:05 master.ldap.test.ed25519_key                                                 
-rw-------. 1 root root  109 Aug 13 02:05 master.ldap.test.ed25519_key.pub                                             
-rw-------. 1 root root 2622 Aug 13 02:05 master.ldap.test.rsa_key                                                     
-rw-------. 1 root root  581 Aug 13 02:05 master.ldap.test.rsa_key.pub                                                 
-rw-------. 1 root root  525 Aug 13 02:05 nfs.test.ecdsa_key                                                           
-rw-------. 1 root root  189 Aug 13 02:05 nfs.test.ecdsa_key.pub                                                       
-rw-------. 1 root root  419 Aug 13 02:05 nfs.test.ed25519_key                                                         
-rw-------. 1 root root  109 Aug 13 02:05 nfs.test.ed25519_key.pub                                                     
-rw-------. 1 root root 2610 Aug 13 02:05 nfs.test.rsa_key
-rw-------. 1 root root  581 Aug 13 02:05 nfs.test.rsa_key.pub   
~~~

pbrezina commented on Makefile:
```diff
@@ -18,13 +18,18 @@ up-keycloak:
 	docker-compose -f docker-compose.yml -f docker-compose.keycloak.yml up \
 	--no-recreate --detach ${LIMIT}
 
+up-ipaipatrust:
```
This is something we can test in PR CI, so I think we can start second IPA with just `make up`.  What do you think?

justin-stephenson replied:
Sure, PR updated.

pbrezina commented on data/configs/dnsmasq.conf:
```diff
@@ -12,6 +12,7 @@ cache-size=0
 
 # These zones have their own DNS server
 server=/ipa.test/172.16.100.10
+server=/ipa2.test/172.16.100.80
```
It does not matter much but you can use `.11` instead of `.80` to keep IPA servers grouped together.

justin-stephenson replied:
Fixed, switched to `172.16.100.11`

pbrezina commented:
> Hi @pbrezina Can you help me understand why `docker.io/ubuntu:rolling` build fails (https://github.com/SSSD/sssd-ci-containers/actions/runs/10370744252/job/28711693110?pr=106) when sshd does not restart properly on `master2.ipa2.test`. I see that the host keys are not found.
> 
> ```
> Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.ecdsa_key                                                    
> Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.ed25519_key                                          
> Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.rsa_key  
> ```
> 
> I ran `src/tools/gen-ssh-keys.sh` and added the ssh keys for `master2.ipa2.test` into this PR. I also added `master2.ipa2.test` into the for loop in `src/tools/gen-ssh-keys.sh`
> 
> However, these are not being copied into the `master2.ipa2.test` system.
> ```

Ubuntu does not provide ipa package so base-ipa container is actually pulled from `quay.io/sssd/ci-base-$svc:lates` which does not contain your changes. Maybe, it would be possible to use the base image we just created? But build.sh would have to change. Maybe introduce UNAVAILABLE_BASE_IMAGE variable or something like that.

pbrezina commented:
> > Hi @pbrezina Can you help me understand why `docker.io/ubuntu:rolling` build fails (https://github.com/SSSD/sssd-ci-containers/actions/runs/10370744252/job/28711693110?pr=106) when sshd does not restart properly on `master2.ipa2.test`. I see that the host keys are not found.
> > ```
> > Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.ecdsa_key                                                    
> > Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.ed25519_key                                          
> > Aug 13 13:19:21 master2.ipa2.test sshd[6905]: Unable to load host key: /data/ssh-keys/hosts/master2.ipa2.test.rsa_key  
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I ran `src/tools/gen-ssh-keys.sh` and added the ssh keys for `master2.ipa2.test` into this PR. I also added `master2.ipa2.test` into the for loop in `src/tools/gen-ssh-keys.sh`
> > However, these are not being copied into the `master2.ipa2.test` system.
> 
> Ubuntu does not provide ipa package so base-ipa container is actually pulled from `quay.io/sssd/ci-base-$svc:lates` which does not contain your changes. Maybe, it would be possible to use the base image we just created? But build.sh would have to change. Maybe introduce UNAVAILABLE_BASE_IMAGE variable or something like that.

Actually it wouldn't work because we run each distro on different host. We would need to store it as artifact and then download it and install it.

justin-stephenson commented:
> > Ubuntu does not provide ipa package so base-ipa container is actually pulled from `quay.io/sssd/ci-base-$svc:lates` which does not contain your changes. Maybe, it would be possible to use the base image we just created? But build.sh would have to change. Maybe introduce UNAVAILABLE_BASE_IMAGE variable or something like that.
> 
> Actually it wouldn't work because we run each distro on different host. We would need to store it as artifact and then download it and install it.

Can ssh keys from both IPA servers `master.ipa.test` and `master2.ipa2.test` be added to `quay.io/sssd/ci-base-ipa:latest` ?

jakub-vavra-cz commented on src/ansible/roles/packages/tasks/Fedora.yml:
```diff
@@ -264,7 +264,7 @@
       - umockdev
     when: passkey_support
-  when: "'base_client' in group_names or 'client' in group_names or 'base_ipa' in group_names or 'ipa' in group_names"
+  when: "'base_client' in group_names or 'client' in group_names or 'base_ipa' in group_names or 'base_ipa2' in group_names or 'ipa' in group_names"
```
Here is still base_ipa2.

justin-stephenson replied:
Removed.

jakub-vavra-cz reviewed (changes requested):
The base_ipa2 is still present.

justin-stephenson commented:
> The base_ipa2 is still present.

Removed fully.

jakub-vavra-cz commented on src/ansible/playbook_image_service.yml:
```diff
@@ -16,7 +16,9 @@
   - samba
 
-- hosts: master.ipa.test
+- hosts:
```
This should target group ipa.

justin-stephenson replied:
Updated as part of rebase.

pbrezina commented:
> > > Ubuntu does not provide ipa package so base-ipa container is actually pulled from `quay.io/sssd/ci-base-$svc:lates` which does not contain your changes. Maybe, it would be possible to use the base image we just created? But build.sh would have to change. Maybe introduce UNAVAILABLE_BASE_IMAGE variable or something like that.
> > 
> > 
> > Actually it wouldn't work because we run each distro on different host. We would need to store it as artifact and then download it and install it.
> 
> Can ssh keys from both IPA servers `master.ipa.test` and `master2.ipa2.test` be added to `quay.io/sssd/ci-base-ipa:latest` ?

No until this PR is merged. But you could do it manually.

pbrezina commented:
Justin, you can try removing the ssh host keys completely when you will rebase on top of Jakub's changes. I'm pretty sure I added them as a workaround for something, but I don't remember anymore. Maybe, it is not needed anymore.

justin-stephenson commented:
> Justin, you can try removing the ssh host keys completely when you will rebase on top of Jakub's changes. I'm pretty sure I added them as a workaround for something, but I don't remember anymore. Maybe, it is not needed anymore.

I rebased and removed the host keys.

jakub-vavra-cz reviewed (approved):
LGTM

danlavu commented:
@justin-stephenson , before I go deeper into the PR. I think we should create a new network for ipa2 and make it more realistic. 

```
server=/ipa2.test/172.16.101.10
```

WDYT?

justin-stephenson commented:
> @justin-stephenson , before I go deeper into the PR. I think we should create a new network for ipa2 and make it more realistic.
> 
> ```
> server=/ipa2.test/172.16.101.10
> ```
> 
> WDYT?

Forgive my ignorance but what is the benefit?

danlavu commented:
> Forgive my ignorance but what is the benefit?

Realism mostly, two domains sharing the same class C network, while it works, it's not suggested  per our installation docs. I don't think it helps us to to test against an unsupported installation.

I do recall there has seem some automation limitations in the past, definitely one, maybe two bugs we would've caught if it wasn't sharing the same subnet.

justin-stephenson commented:
> > Forgive my ignorance but what is the benefit?
> 
> Realism mostly, two domains sharing the same class C network, while it works, it's not suggested per our installation docs. I don't think it helps us to to test against an unsupported installation.
> 
> I do recall there has seem some automation limitations in the past, definitely one, maybe two bugs we would've caught if it wasn't sharing the same subnet.

I don't have a strong opinion either way, @pbrezina can you provide your thoughts?

danlavu commented:
> I don't have a strong opinion either way, @pbrezina can you provide your thoughts?

FWIW, I feel strongly about this. Testing against a supported/gss certified setup than a test. While it doesn't apply here, the downstream multi-domain AD stuff needs to be segmented this way otherwise dcpromo will fail.

pbrezina commented:
> > > Forgive my ignorance but what is the benefit?
> > 
> > 
> > Realism mostly, two domains sharing the same class C network, while it works, it's not suggested per our installation docs. I don't think it helps us to to test against an unsupported installation.
> > I do recall there has seem some automation limitations in the past, definitely one, maybe two bugs we would've caught if it wasn't sharing the same subnet.
> 
> I don't have a strong opinion either way, @pbrezina can you provide your thoughts?

I think it just unnecessarily complicates the setup. I have no idea if adding a network in a docker-compose is sufficient or additional tinkering to make the networks accessible from each other will be required.

But I don't have opinion whether it should or should not be done, I'm fine either way.

danlavu commented:
A new docker net will be sufficient, I know it's extra complexity, but it's worth it.

justin-stephenson commented:
> A new docker net will be sufficient, I know it's extra complexity, but it's worth it.

Is it more than a simple one-line change? If yes would you mind adding a commit to this PR with these changes and pushing to this PR?

jakub-vavra-cz commented:
Just be aware that we can't copy this that easily in idmci (without hadcoding specific openstack networks). We also  need to make sure that the same network will still work.

danlavu commented:
> Just be aware that we can't copy this that easily in idmci (without hadcoding specific openstack networks). We also need to make sure that the same network will still work.

Ack, I did run it in IDMCI and the provision passes, I'll go ahead and test it. 

```
2024-12-10T05:40:24 PLAY RECAP *********************************************************************
2024-12-10T05:40:24 client.test                : ok=82   changed=46   unreachable=0    failed=0    skipped=45   rescued=0    ignored=0   
2024-12-10T05:40:24 dc-nhhq.ad-nhhq.test       : ok=17   changed=8    unreachable=0    failed=0    skipped=8    rescued=0    ignored=0   
2024-12-10T05:40:24 dc.samba.test              : ok=67   changed=44   unreachable=0    failed=0    skipped=44   rescued=0    ignored=0   
2024-12-10T05:40:24 dns.test                   : ok=57   changed=26   unreachable=0    failed=0    skipped=42   rescued=0    ignored=0   
2024-12-10T05:40:24 kdc.test                   : ok=57   changed=27   unreachable=0    failed=0    skipped=46   rescued=0    ignored=0   
2024-12-10T05:40:24 master.ipa.test            : ok=61   changed=35   unreachable=0    failed=0    skipped=47   rescued=0    ignored=0   
2024-12-10T05:40:24 master.ldap.test           : ok=61   changed=31   unreachable=0    failed=0    skipped=44   rescued=0    ignored=0   
2024-12-10T05:40:24 master2.ipa2.test          : ok=60   changed=34   unreachable=0    failed=0    skipped=48   rescued=0    ignored=0   
2024-12-10T05:40:24 nfs.test                   : ok=54   changed=24   unreachable=0    failed=0    skipped=46   rescued=0    ignored=0  
```

justin-stephenson commented:
I rebased this PR, and also added capabilities to the 'ipa2' section in docker-compose.yml similar to

https://github.com/SSSD/sssd-ci-containers/commit/211908d635758488211803530151aa43d6b04048

danlavu commented:
@justin-stephenson, I have the configuration provisioning but the networks are not routable to one another so I'm trying to figure that out.

sumit-bose commented:
Hi,

jfyi, I was still able to setup a test environment with this PR.

bye,
Sumit

danlavu commented:
We have to visit the multiple network stuff later. To @jakub-vavra-cz  point, the priority is that it works on the same net.

danlavu commented:
We have to visit the multiple network stuff later. To @jakub-vavra-cz  point, the priority is that it works on the same net.

justin-stephenson commented:
Hi @pbrezina Do you think this is ready to be merged?

Related PRs https://github.com/SSSD/sssd-test-framework/pull/119 and https://github.com/SSSD/sssd/pull/7517 will still be blocked until FreeIPA IDM IDM Trust code is merged.

Based on your earlier comment:

> Justin, you can try removing the ssh host keys completely when you will rebase on top of Jakub's changes. I'm pretty sure I added them as a workaround for something, but I don't remember anymore. Maybe, it is not needed anymore.

I created https://github.com/SSSD/sssd-ci-containers/pull/122 for this but please check if I missed anything.

pbrezina commented on data/configs/dnsmasq.conf:
```diff
@@ -35,3 +36,4 @@ ptr-record=30.100.16.172.in-addr.arpa,dc.samba.test
 ptr-record=40.100.16.172.in-addr.arpa,client.test
 ptr-record=10.200.16.172.in-addr.arpa,dc.ad.test
 ptr-record=70.100.16.172.in-addr.arpa,master.keycloak.test
+ptr-record=80.100.16.172.in-addr.arpa,master2.ipa2.test
```
This can be `master.ipa2.test` as it is the only server in the second IPA domain.

justin-stephenson replied:
Updated.

pbrezina reviewed (changes requested):
Hi, see a nitpick inline. Otherwise it looks good.

justin-stephenson commented:
> Hi, see a nitpick inline. Otherwise it looks good.

@pbrezina friendly ping reminder

pbrezina commented:
Ack, but I am rerunning c8s before pushing as there is a failure.

justin-stephenson commented:
> Ack, but I am rerunning c8s before pushing as there is a failure.

No host keys were added to `data/ssh-keys/hosts/` for `master.ipa2.test` in this PR, therefore the below step fails because c8s openssh version does not appear to support this drop directory `/etc/ssh/sshd_config.d/`. How do you prefer to fix it?
```

- name: Configure SSH daemon without pre-generated hostkey
  ansible.builtin.copy:
    dest: /etc/ssh/sshd_config.d/sshd.conf
    owner: root
    group: root
    mode: 0600
    content: |
      PermitRootLogin yes
  when: not stat_ecdsa_key.stat.exists

```

pbrezina commented:
The problem is not that sshd_config.d is not supported on C8S, but `copy` does not create the folder, while `template` which is used to set the generated keys does.

danlavu commented:
Last week, I tried to get the networks working. Everything is provisioned, but the hosts are not routable to one another. Do you think you can test it? Perhaps it's an issue with my workstation? 

https://github.com/justin-stephenson/sssd-ci-containers/pull/1

I suspect that it may be a firewall-related issue. I've tried the following with no such luck. 
```
firewall-cmd --direct --add-rule ipv4 filter FORWARD 0 -i podman1 -o podman2  -j ACCEPT
firewall-cmd --direct --add-rule ipv4 filter FORWARD 0 -i podman2 -o podman1  -j ACCEPT
```

Then, explicitly defined the networks

```
firewall-cmd --zone=FedoraWorkstation --add-rich-rule='rule family="ipv4" source address="172.16.100.0/24" destination address="172.16.110.0/24" accept'
firewall-cmd --zone=FedoraWorkstation --add-rich-rule='rule family="ipv4" source address="172.16.110.0/24" destination address="172.16.100.0/24" accept'
```
